src/tactic/has_variable_names.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2020 Jannis Limperg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jannis Limperg
+-/
+
+import tactic.core
+
+/-!
+# A tactic for type-based naming of variables
+
+When we name hypotheses or variables, we often do so in a type-directed fashion:
+a hypothesis of type `ℕ` is called `n` or `m`; a hypothesis of type `list ℕ` is
+called `ns`; etc. This module provides a tactic,
+`tactic.typical_variable_names`, which looks up typical variable names for a
+given type. Typical variable names are registered by giving an instance of the
+type class `has_variable_names`.
+-/
+
+universes u v
+
+/--
+Type class for associating a type `α` with typical variable names for elements
+of `α`. See `tactic.typical_variable_names`.
+-/
+meta class has_variable_names (α : Sort u) : Type :=
+(first_name : name)
+(other_names : list name)
```

Comment:
Is the reason for having two fields just that the list has to be nonempty? Is that prefereable over a list + proof for `0 < list.length`?

Reply:
Yes, this is just the low-tech definition of a nonempty list. If you prefer the subtype, that's also fine with me, but I don't see how this would make constructing or consuming instances easier.

Reply:
Is there a reason to require non-emptiness at all?  There needs to be fallback code in case there are no typical variables anyhow.  Also this is meta code, so it would even be ok to fail.  Besides, ``[`m, `n]`` is simply much easier to read than `⟨`m, [`n]⟩`.

Reply:
The non-emptiness requirement is there because I can see no reason to have an empty instance. Without it, we would have to check in `typical_variable_names` whether the provided list is indeed nonempty. I generally favour enforcing such invariants as early as possible and in the types if this can be done at low cost. But the impact of all this is marginal, so if you want the class with an implicit invariant, I'll implement that as well.

Reply:
I just really dislike ``⟨`m, [`n]⟩``.  How about this?
```lean
class has_var_names (α : Sort u) : Type :=
(names : list name)
(neq_nil : names ≠ [] . tactic.exact_dec_trivial)

meta instance : has_var_names ℕ := ⟨[`m, `n]⟩
meta instance : has_var_names string := ⟨[`s, `t]⟩
```
(I still think the nonemptiness requirement is overkill, though.)

Reply:
Looks good to me. I'll implement this version.